crucible-wasm/src/Lang/Crucible/Wasm/Instantiate.hs

```diff
@@ -0,0 +1,531 @@
+{-# LANGUAGE BlockArguments #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE ParallelListComp #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE ViewPatterns #-}
+
+module Lang.Crucible.Wasm.Instantiate where
+
+import Control.Exception
+import Control.Monad
+import Control.Monad.Trans
+import Control.Monad.Trans.State
+import Control.Monad.Trans.Except
+import Numeric.Natural
+import GHC.Stack
+
+import qualified Data.ByteString.Lazy as LBS
+import qualified Data.Text as T
+import qualified Data.Text.Lazy as TL
+import qualified Data.Map as Map
+import qualified Data.IntMap as IntMap
+import qualified Data.Sequence as Seq
+import           Data.String
+import           Data.Word
+
+import Data.Parameterized.Context
+import Data.Parameterized.Some
+
+import qualified Language.Wasm as Wasm
+import qualified Language.Wasm.Structure as Wasm
+import qualified Language.Wasm.Validate as Wasm
+
+import Lang.Crucible.Simulator
+import Lang.Crucible.CFG.Common (freshGlobalVar)
+import Lang.Crucible.FunctionHandle
+import Lang.Crucible.Types
+
+import Lang.Crucible.Wasm.Memory( WasmMem )
+
+
+defineModule ::
+  HandleAllocator ->
+  Namespace ->
+  Wasm.ModuleDef ->
+  InstM (Maybe Wasm.Ident, Wasm.ValidModule, ModuleInstance)
+defineModule halloc ns mdef =
+  do (ident,m) <- buildModule mdef
+     vm <- case Wasm.validate m of
+             Left verr -> throwE (ValidationError verr)
+             Right vm  -> pure vm
+     im <- instantiateModule halloc ns vm
+     pure (ident, vm, im)
+
+
+buildModule :: Wasm.ModuleDef -> InstM (Maybe Wasm.Ident, Wasm.Module)
+buildModule (Wasm.RawModDef ident m) = pure (ident, m)
+buildModule _ = bailOut "build module"
```

Comment:
Just curious if the semantics here are different than for `panic`?

Reply:
And if so, maybe document why.

Reply:
I mostly used this for unimplemented stuff, which I'm starting to think should be different from `panic` as I discussed above.